System/Console/ANSI/Color.hs

```diff
@@ -75,46 +85,85 @@ clamp n min_ max_
+-- | a Color8Code represents the escape code for an 8-bit ANSI Color. It is in
+-- range [0..255]. The ranges of its values are:
+-- (from https://en.wikipedia.org/wiki/ANSI_escape_code#Colors)
 --
 -- 0x00-0x07:  standard colors (as in ESC [ 30–37 m)
 -- 0x08-0x0F:  high intensity colors (as in ESC [ 90–97 m)
--- 0x10-0xE7:  6 × 6 × 6 cube (216 colors): 16 + 36 × r + 6 × g + b (0 ≤ r, g, b ≤ 5)
+-- 0x10-0xE7:  6 × 6 × 6 cube (216 colors):
+--             16 + 36 × r + 6 × g + b (0 ≤ r, g, b ≤ 5)
 -- 0xE8-0xFF:  grayscale from black to white in 24 steps
 --
--- The colors described by the 2 first ranges (0x00-0x07 and 0x08-0x0F) match exactly with
--- the 4 bit ANSI colors represented by (ColorIntensity, Color).
--- The colors of the 3rd range (0x10-0xE7) are represented by RGB8Color, and the colors of the 4th range
--- are represented by Gray8Color.
-color8CodeToEither :: Color8Code
-                   -> Either (ColorIntensity, Color) Color8
-color8CodeToEither (Color8Code c)
-  | c < 16    = Left $ fst $ aNSIColors !! fromIntegral c   -- interpreted as a 4-bit ANSI color
-  | c < 232   = Right $ asRGB (c-16)          -- interpreted as 8-bit rgb
-  | otherwise = Right $ Gray8Color (c-232)    -- interpreted as 8-bit grayscale
+-- The colors described by the 2 first ranges (0x00-0x07 and 0x08-0x0F) match
+-- exactly with the 4 bit ANSI colors represented by (ColorIntensity, Color).
+-- The colors of the 3rd range (0x10-0xE7) are represented by RGB8Color, and the
+-- colors of the 4th range are represented by Gray8Color.
+color8CodeToXterm256 :: Color8Code -> Xterm256Color
+color8CodeToXterm256 (Color8Code c)
+  | c < 16    = SystemColor intensity color -- interpreted as a 4-bit ANSI color
+  | c < 232   = RGBColor $ asRGB (c - 16)   -- interpreted as 8-bit rgb
+  | otherwise = GrayColor (c - 232)         -- interpreted as 8-bit grayscale
  where
-  asRGB i = let -- we know that i = 36 × r + 6 × g + b and (0 ≤ r, g, b ≤ 5) (cf. comment on top)
-                -- so we can deduce the unique set of corresponding r g and b values:
+  (intensity, color) = fst $ aNSIColors !! fromIntegral c
+  asRGB i = let -- we know that i = 36 × r + 6 × g + b and (0 ≤ r, g, b ≤ 5)
+                -- (cf. comment on top) so we can deduce the unique set of
+                -- corresponding r g and b values:
                 r = quot i 36
                 g = quot (r - 36*i) 6
                 b = i - (6*g + 36*i)
-            in  RGB8Color r g b
+            in  RGB r g b
+
+-- | Color8 represents an 8-bit ANSI color. This function converts it to the
+-- corresponding code, defined by https://en.wikipedia.org/wiki/ANSI_escape_code#Colors
+--
+-- For safety the values of Color8 are clamped in their respective ranges.
+xterm256ColorToCode :: Xterm256Color -> Color8Code
+xterm256ColorToCode (SystemColor Dull color)
+  = Color8Code (fromIntegral $ colorToCode color)
+xterm256ColorToCode (SystemColor Vivid color)
+  = Color8Code (fromIntegral $ 16 + colorToCode color)
```

Comment:
I think it's 8 that should be added instead of 16?

Reply:
Agreed - thank you for spotting my error.